dataset.go

```diff
@@ -86,6 +114,70 @@ func flatElementsIterator(elems []*Element, elemChan chan<- *Element) {
 	}
 }
 
+// FlatDatasetIterator represents a simple iterator over the elements in a
+// Dataset.
+type FlatDatasetIterator interface {
+	// HasNext indicates if the iterator as another element.
+	HasNext() bool
+	// Next gets and returns the next element in the iterator.
+	Next() *Element
+}
+
+// FlatStatefulIterator returns a stateful iterator that adheres to
+// FlatDatasetIterator interface. This allows the caller to iterate over every
+// element in the dataset, including elements nested inside sequences.
+//
+// Important note: if the Dataset changes during the iteration (e.g. if elements
+// are added or removed), those elements will not be included until a new
+// iterator is created.
+//
+// If you don't need to receive elements on a/ channel, and don't want to worry
+// about always exhausting this iterator, this is the best and safest way to
+// iterate over a Dataset.
+func (d *Dataset) FlatStatefulIterator() FlatDatasetIterator {
+	flatElems := flatElemsWrapper{}
+	flatSliceBuilder(d.Elements, &flatElems)
+	// fmt.Println("Dataset:", flatElems.flatElems)
+	iter := flatIterator{flattenedDataset: flatElems.flatElems}
+	return &iter
```

Comment:
This is a nit, but I think you could return a non-pointer here. The struct only has an int and a slice, so it's not a lot of memory to copy around, and you could avoid a pointer-chase on calling HasNext()

Reply:
Typically, folks use pointer receivers on struct methods that modify the struct. This iterator is stateful, e.g. `Next()` modifies the struct itself. Because the methods are defined on the pointer receiver, we should return the pointer to the caller.